test/FunctionalTests/Microsoft.Extensions.OData.Migration.Tests/UriTranslationTests.cs

```diff
@@ -0,0 +1,170 @@
+﻿using System;
+using System.IO;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+
+namespace Microsoft.Extensions.OData.Migration.Tests
+{
+    [TestClass]
+    public class UriTranslationTests
+    {
+        private Uri serviceRoot;
+        private TranslationMiddleware middleware;
+
+        [TestInitialize]
+        public void Initialize()
+        {
+            string v4edmx = SampleModelBuilder.TransformODataV3EdmxToODataV4Edmx("V3SampleService.xml", "V2-to-V4-CSDL-NoSap.xsl");
+            var v3model = SampleModelBuilder.LoadTestv3Model("V3SampleService.xml");
+            var v4model = SampleModelBuilder.LoadTestv4Model(v4edmx);
+
+            // for actual translation, root is irrelevant.  Root is however relevant for determining what part of uri should be translated
+            this.serviceRoot = new Uri("http://foo:80/odata");
+            this.middleware = new TranslationMiddleware(v3model, v4model, new MigrationOptions()
+            {
+                ServiceRoot = serviceRoot
+            });
+        }
+
+        /// TEST TODOS
+        /// NavigationPropertyLink tests (need clarification)
+        /// BatchReferenceLink tests (need clarification)
+        /// Type Segment (need clarification)
+        /// Bad models and exception throwing
+
+        #region EntitySet URI Translation Tests
+        [TestMethod]
+        public void TestEntitySet()
+        {
+            Assert.AreEqual(serviceRoot.Append("/Persons"), middleware.TranslateUri(serviceRoot.Append("/Persons")));
+        }
+
+        [TestMethod]
+        public void TestEntitySetNotFoundInV3Model ()
+        {
+            Assert.ThrowsException <Data.OData.Query.ODataUnrecognizedPathException>(() =>
+            {
+                middleware.TranslateUri(serviceRoot.Append("/doesntexist"));
+            });
+        }
+
+        [TestMethod]
+        public void TestEntitySet2()
+        {
+            Assert.AreEqual(serviceRoot.Append("/Products"), middleware.TranslateUri(serviceRoot.Append("/Products")));
+        }
+        #endregion
+
+        #region NavigationProperty/Link URI Translation Tests
+        [TestMethod]
+        public void TestCollectionNavigationProperty ()

```

Comment:
My opinion on test methods is that they can be long (if needed) to clearly describe what you are testing. I can then look at the name of the test and see what you are testing. I might call this WhenV3CollectionIsSpecifiedItCorrectlyGetsTranslatedToV4Collection -- or something like that. Nobody ever calls these methods except the test framework so verbose names is not a bad thing.